lib/rest-simple-interface.js

```diff
@@ -0,0 +1,298 @@
+// Copyright 2016 Zipscene, LLC
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+const express = require('express');
+const bodyParser = require('body-parser');
+const _ = require('lodash');
+const XError = require('xerror');
+const APIInterfaceJSONBase = require('./api-interface-json-base');
+const APIInterface = require('./api-interface');
+const utils = require('./utils');
+const zstreams = require('zstreams');
+const pasync = require('pasync');
+const KeepAlive = require('./keep-alive');
+const CrispHooks = require('crisphooks');
+const objtools = require('objtools');
+
+/**
+ * API interface for REST.
+ *
+ * @class RESTSimpleInterface
+ * @constructor
+ * @extends APIInterfaceJSONBase
+ * @param {Object} [options={}]
+ *   @param {Boolean} [options.includeErrorStack=false] - Whether to include stack traces in error responses.
+ * @since v1.0.0
+ */
+class RESTSimpleInterface extends APIInterfaceJSONBase {
+
+	constructor(options = {}) {
+		super(options);
+
+		// Map of registered methods
+		this.methods = {};
+	}
+
+	/**
+	 * Given an Express router, registers this interface to handle its portion of API calls from the router.
+	 * See parent class for detailed documentation
+	 *
+	 * @method registerInterfaceWithRouter
+	 * @param {express.Router} router
+	 * @since v1.0.0
+	 */
+	registerInterfaceWithRouter(router) {
+		router.use('/rest', this.router);
+	}
+
+	sendHeader(ctx) {
+		// Send headers
+		let contentType = (ctx.routeOptions.streamingResponse && !ctx.error) ? 'text/plain' : 'application/json';
+		let statusCode = 200;
+		if (ctx.error) {
+			if (XError.isXError(ctx.error)) {
+				statusCode = (XError.getErrorCode(ctx.error.code) || {}).http || 500;
+			} else {
+				statusCode = 500;
+			}
+		}
+		ctx.res.writeHead(statusCode, {
+			'Content-type': `${contentType}; charset=utf-8`
+		});
+	}
+
+	/**
+	 * This is called for any REST request.
+	 */
+	_mainRESTMiddleware(req, res, next) {
+		let pathComponents = req.path.replace(/^\//, '').replace(/\/\//g, '/').split('/');
+		let body = (req.body && _.isPlainObject(req.body)) ? req.body : {};
+
+		const getQueryParam = (name) => {
+			let str = req.query && req.query[name];
+			if (str) {
+				try {
+					return JSON.parse(str);
+				} catch (err) {
+					return str;
+				}
+			} else {
+				return undefined;
+			}
+		};
+
+		const findObjectOperationMethod = (methodSuffix) => {
```

Comment:
This is gross and I hate REST. That is all

Reply:
Yup